Sources/SwiftDocCTestUtilities/XCTestCase+TemporaryDirectory.swift

```diff
@@ -0,0 +1,62 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import XCTest
+
+// These helpers methods exist to put temp files for different test executions in different locations when running in Swift CI.
+
+public extension XCTestCase {
+    
+    @available(*, deprecated, message: "Use `createTemporaryDirectory` instead in unit tests to avoid referencing a shared location in Swift CI.")
+    func NSTemporaryDirectory() -> String {
+        return Foundation.NSTemporaryDirectory()
+    }
+    
+    /// Creates a new temporary directory and returns the URL of that directory.
+    ///
+    /// At the end of the test the temporary directory is automatically removed.
+    ///
+    /// - Parameters:
+    ///   - pathComponents: Additional path components to add to the temporary URL.
+    ///   - createDirectoryForLastPathComponent: If the file manager should create a directory for the last path component or not. Defaults to `true`.
+    ///   - fileManager: The file manager that will create the directory.
+    /// - Returns: The URL of the newly created directory.
+    func createTemporaryDirectory(
+        pathComponents: String...,
+        createDirectoryForLastPathComponent: Bool = true,
```

Comment:
I think this parameter could lead to confusion especially in the case where no path components are specified:

```swift
createTemporaryDirectory(createDirectoryForLastPathComponent: false)
```

It's not clear at the call site what the "last path component" means in this case. 

It seems like this is something we need because the `TempFolder.write(to:)` is responsible for creating the directory at the moment? I wonder if we should make `TempFolder` call `createTemporaryDirectory` directly and not create the directory in `write`. That way, we don't need to pass an `atRoot` argument. Another way would be to create a `createTempFolder` method in this extension that returns a `TempFolder`, which would handle the internals of creating the folder at the temporary location. Thoughts?

Reply:
Also, the case where there are path components, such as:

```swift
try createTemporaryDirectory(pathComponents: "test.docc", createDirectoryForLastPathComponent: false)
```

might be better written as:

```swift
try createTemporaryDirectory().appendingPathComponent("test.docc")
```

Reply:
> I wonder if we should make TempFolder call createTemporaryDirectory directly and not create the directory in write. That way, we don't need to pass an atRoot argument. Another way would be to create a createTempFolder method in this extension that returns a TempFolder, which would handle the internals of creating the folder at the temporary location. Thoughts?

With how the code is currently structured this isn't possible. `TempFolder` is defined internally to the `SwiftDocC` target and this can't be imported here. We don't want to make it public API and we also can't import `XCTest` in `SwiftDocC` to reference `XCTestCase` there. Longer term we should probably move both the `File` and `Folder` types, the `TestFileSystem`, and a few more related things into the test utilities target but I'm not sure if we should do that work in this PR or not.

Reply:
I've removed the `createDirectoryForLastPathComponent` argument here.